src/LoadingSwitch.jsx

```diff
@@ -1,123 +0,0 @@
-    children,
-    loadingScreen: LoadingScreen,
-    maxLoadingTime = 0
-}) => {
-    const context = useContext(RouterContext);
-    const loadingContext = useContext(LoadingContext);
-    const isCurrentlyLoading = useContext(LoadingGetterContext);
-
-    const [current, setCurrent] = useState(() => {
-        const isFirstPageLoadable = isLoadable(context, children);
-
-        // if first page uses loading then show loading screen
-        return isFirstPageLoadable
-            ? { ...context, location: { pathname: loadingPathname } }
-            : context;
-    });
-    const [next, setNext] = useState(current);
-
-    const timeout = useRef();
-
-    // when location changed
-    useEffect(() => {
-        // if not the same route mount it to start loading
-        if (isPathsDifferent(context, next)) {
-            const isPageLoadable = isLoadable(context, children);
-
-            setNext({ ...context });
-
-            if (!isPageLoadable) {
-                loadingContext.done();
-                setCurrent({ ...context });
-            } else {
-                if (!isCurrentlyLoading)
-                    loadingContext.start();
-                else
-                    loadingContext.restart();
-            }
-        }
-
-        // if same as current route stop loading
-        if (isPathsEqual(context, current)) {
-            loadingContext.done();
-
-            if (isSearchDifferent(context, current))
-                setCurrent({ ...context });
-        }
-    }, [context]);
-
-    // when loading ends
-    useEffect(() => {
-        if (!isCurrentlyLoading && isPathsDifferent(current, next))
-            setCurrent(next);
-    }, [isCurrentlyLoading]);
-
-    // setTimeout if maxLoadingTime is provided
-    useEffect(() => {
-        if (maxLoadingTime > 0) {
-            if (timeout.current) {
-                clearTimeout(timeout.current);
-                timeout.current = undefined;
-            }
-
-            if (isPathsDifferent(current, next)) {
-                timeout.current = setTimeout(() => {
-                    loadingContext.done();
-                }, maxLoadingTime);
-            }
-        }
-    }, [current, next]);
-
-    // memo current and next components
-    return useMemo(
-        () => <>
-            {/* current */}
-            {
-                current.location.pathname !== loadingPathname
-                    ? <RouteComponent key={current.location?.pathname} context={current} allRoutes={children} />
-                    : LoadingScreen
-                        ? <LoadingScreen location={current.location} />
-                        : <DefaultLoadingScreen location={current.location} />
-            }
-
-            {/* hidden next */}
-            {
-                isPathsDifferent(current, next) &&
-                <RouteComponent key={next.location?.pathname} context={next} allRoutes={children} hidden />
-            }
-        </>,
-        [current, next]
-    );
```

Comment:
Here I load the `current` route and the hidden `next` one while the `next` one is loading. Haven't found something like that in your solution. 

Also, I have to say that it's really hard to review both support of react-router 6 and migration to TS at the same time. It would be much better to do it separately. I will probably take your code as a reference and do it in two steps if you don't mind

Reply:
Yeah, I dropped that part and then realized that it wasn't working properly without it, so I restored it back the next day ) just forgot to push it to this PR. Works now. Thanks for the great concept! React Router 6 has made many breaking changes, so I had to workaround it a bit.